#11 Port from rx to coroutines

victorvicari opened the pull request:
Some dependencies were added (coroutines, test, work manager coroutines test...). Basically leaving the project structure as it was, and I moved from rx to coroutines. About the tests, I used MockK once that it's better to deal with the kotlin coroutines. Some issues came up when I updated the workmanager, I solved by putting the java `source` and `target` at the `gradle` and `kotlinOption`

victorvicari committed:
updated gradle version to 3.4.1.
added the properties trim and capitalize to every partner name

victorvicari committed:
sorted collaborators list with lower-case

victorvicari committed:
replaced rxJava with Kotlin Coroutines.
updated some gradle dependencies

victorvicari committed:
view model test with coroutines

victorvicari committed:
added mockk to unit test
bug fix at EventsViewModel loading progress

victorvicari committed:
loading bug fix

victorvicari committed:
updated refreshEventsJob unit test to work with coroutines

victorvicari committed:
updated some gradle dependencies. also added some for testing purpose

victorvicari committed:
added the coroutines impl and was necessary to jvmTarget and javaTarget because of the workmanager version

victorvicari committed:
minor bug fix

victorvicari committed:
fix

ashdavies commented on app/build.gradle:
```diff
@@ -24,17 +24,30 @@ android {
+    // Coroutines With Arch Components
+    implementation("androidx.lifecycle:lifecycle-viewmodel-ktx:${rootProject.archCompCoroutinesVersion}")
+    implementation("androidx.lifecycle:lifecycle-runtime-ktx:${rootProject.archCompCoroutinesVersion}")
+    implementation("androidx.lifecycle:lifecycle-livedata-ktx:${rootProject.archCompCoroutinesVersion}")
```
Use `rootProject.lifecycleVersion` for consistency with already imported `lifecycle-extensions`, additionally using an alpha library may bring stability issues when a stable library was already in use.

victorvicari replied:
Ok.

ashdavies commented on app/build.gradle:
```diff
@@ -50,6 +63,8 @@ dependencies {
 
     // Mockito
     testImplementation("org.mockito:mockito-core:${rootProject.mockitoVersion}")
+    testImplementation("io.mockk:mockk:${rootProject.mockkVersion}")
```
If Mockito is already in use, I would really recommend using `mockito-kotlin` as it’s a bridge between the two rather than introducing a whole new mocking framework.

victorvicari replied:
I understand

ashdavies commented on app/src/main/java/com/wtmberlin/data/ApiService.kt:
```diff
@@ -3,53 +3,65 @@ package com.wtmberlin.data
 open class ApiService(private val meetupService: MeetupService) {
-    fun events(): Single<List<WtmEvent>> {
-        return meetupService.events().map { it.map(MeetupEvent::toWtmEvent) }
+    suspend fun events() = withContext(Dispatchers.IO) {
```
You probably shouldn’t immediately specify a dispatcher within a suspended method call, not least since it overrides the expected behaviour, but it becomes very difficult to test as the dispatcher is not injected.

ashdavies commented on app/src/main/java/com/wtmberlin/data/ApiService.kt:
```diff
@@ -3,53 +3,65 @@ package com.wtmberlin.data
+            }
+        }.toString()
+    }
+}
```
nit: missing new line

ashdavies commented on app/src/main/res/layout/events_screen.xml:
```diff
@@ -13,10 +13,10 @@
+            android:id="@+id/swipe_refresh"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            app:refreshing="@{safeUnbox(viewModel.refreshing)}">
```
Does this project use `safeUnbox`? This seems unrelated to this PR.

ashdavies commented on app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt:
```diff
@@ -1,107 +1,162 @@
-import com.wtmberlin.data.Result
-import com.wtmberlin.data.WtmEvent
-import io.reactivex.processors.PublishProcessor
+import com.wtmberlin.data.*
```
nit: avoid wildcard imports

ashdavies commented on app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt:
```diff
@@ -1,107 +1,162 @@
-    }
+    @MockK
+    internal lateinit var repository: Repository
+    private val viewModel by lazy { EventDetailsViewModel("1", repository) }
```
Use `LazyThreadSafetyMode.NONE` with the `lazy` delegate to avoid having the performance cost of a synchronised double check when accessing in tests or in single threaded environments.

victorvicari replied:
Ok

ashdavies commented on app/src/test/java/com/wtmberlin/works/RefreshEventsJobTest.kt:
```diff
@@ -1,67 +1,80 @@
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+        Dispatchers.setMain(testDispatcher)
```
Consider creating a `Rule` using a `TestDispatcher` where you don't need to manipulate it under test

victorvicari replied:
Done

ashdavies commented on app/src/test/java/com/wtmberlin/works/RefreshEventsJobTest.kt:
```diff
@@ -1,67 +1,80 @@
-        worker.doWork()
+        work.doWork()
 
-        verify(eventDao).replaceAll(listOf(eventFromNetwork))
```
It really doesn’t seem that necessary to migrate the mocking framework to something else when `Mockito` is already in use.

ashdavies reviewed (changes requested):
Thanks for the contribution @victorvicari , I have a few comments regarding your pull request.

- Verify that chaning the JVM target from 7 to 8 doesn’t introducing any regression issues
- Inject `Dispatcher`’s instead of hard coding, and only apply then inside a coroutine context
- I don’t think it’s appropriate to introduce a new mock framework when one is already in use
- Use lifecycle `2.1.0-rc01` instead of alpha to make use of `viewModelScope` instead of alpha

You can find a good example of Retrofit, Coroutine and Dispatcher use with the Plaid project.
https://github.com/android/plaid/blob/885a002a4f7f56ed7ff8c69c7866957591727a4c/core/src/main/java/io/plaidapp/core/data/CoroutinesDispatcherProvider.kt

victorvicari committed:
added some dependencies for testing purpose

victorvicari committed:
removed safeUnbox

victorvicari committed:
Like plaid, injecting the coroutines into the constructor.
Updated result, to avid try catch inside the code

victorvicari committed:
minor refactor into the view layer.
added a new class, to added there some crash library.

victorvicari committed:
Base viewmodel with coroutines and error logger class

victorvicari committed:
removed mockk in those tests class. replaced by mockito-kotlin

ashdavies commented on app/src/test/java/com/wtmberlin/ui/EventsViewModelTest.kt:
```diff
@@ -1,92 +1,126 @@
-        every { it.meetupUrl } returns "www.google.com"
-    }
+    private fun startEvents(event: WtmEvent = mock()): WtmEvent {
+        Mockito.`when`(event.id).thenReturn("1")
```
You can also use the `given` … `when` … `then` syntax from `BDDMockito` to avoid protected keywords in Kotlin

victorvicari replied:
Done. Thanks for the tip

ashdavies reviewed (approved):
Great improvement! Thanks for putting the effort in  👍 💯

victorvicari commented:
Still missing some things. I'll try to finish today.

victorvicari committed:
I created a rule for dispatchers tests. Also updated the RefreshEventsJobTest

ashdavies commented on app/src/test/java/com/wtmberlin/SetMainDispatcherRule.kt:
```diff
@@ -0,0 +1,22 @@
+class SetMainDispatcherRule : TestWatcher() {
+
+    override fun starting(description: Description?) {
+        Dispatchers.setMain(newSingleThreadContext("UI thread"))
```
Don’t forget you’ll need to release this thread on test finish

faogustavo replied:
But he is already doing this

ashdavies replied:
https://github.com/Kotlin/kotlinx.coroutines/tree/master/kotlinx-coroutines-test

Check the example where the surrogate thread is being closed.

faogustavo commented on app/build.gradle:
```diff
@@ -43,13 +55,20 @@ dependencies {
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:${rootProject.coroutinesTestCoreVersion}")
+    testImplementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:${rootProject.coroutinesTestCoreVersion}")
+    testImplementation("org.jetbrains.kotlinx:kotlinx-coroutines-test:${rootProject.coroutinesTestVersion}")
+    testImplementation("com.nhaarman.mockitokotlin2:mockito-kotlin:${rootProject.mockitoKotlinVersion}")
```
Can you split the `implementation` from `testImplementation` to make it a little bit more organized

victorvicari replied:
Done

faogustavo commented on app/src/main/java/com/wtmberlin/data/Repository.kt:
```diff
@@ -1,67 +1,56 @@
 
-    fun refreshEvents() {
-        eventsResource.refresh()
+    suspend fun refreshEvents() = withContext(dispatchers.io) {
```
You can simplify it by doing this
```suggestion
    suspend fun refreshEvents() = withContext(dispatchers.io, ::fetchEvents)
```

faogustavo commented on app/src/main/java/com/wtmberlin/data/WtmEventDao.kt:
```diff
@@ -4,15 +4,14 @@ import androidx.room.Dao
 
     @Query("SELECT * FROM WtmEvent WHERE id = :eventId")
-    fun getById(eventId: String): Flowable<WtmEvent>
+    suspend fun getById(eventId: String): WtmEvent
```
You should use an optional here, to avoid crashing the app if the item doesn't exist.
```suggestion
    suspend fun getById(eventId: String): WtmEvent?
```

victorvicari replied:
Ok. I'll add

faogustavo commented on app/src/main/java/com/wtmberlin/util/ErrorLogger.kt:
```diff
@@ -0,0 +1,10 @@
+
+class ErrorLogger {
+
+    fun getException(e: Throwable) {
```
Maybe you can choose a meaningful name, like `logException`

victorvicari replied:
Done

faogustavo commented on app/src/main/java/com/wtmberlin/util/ErrorLogger.kt:
```diff
@@ -0,0 +1,10 @@
+
+    fun getException(e: Throwable) {
+        //Todo: add a crash lib
+        val a = e
```
I think is missing the Timber rule here. It will be nice to also check if its a prod build and avoid logging with timber and then, maybe, add a crash lib like `crashlytics` or `sentry`.

victorvicari replied:
What do you think @ashdavies

ashdavies replied:
Conditional reporting can be achieved with a debug `Timber.Tree` which can always be beneficial, but crash reporting should be configured by the project owner since it requires setup of an additional service and an API key which can sometimes incur billing.

faogustavo commented on app/src/test/java/com/wtmberlin/ui/EventsDetailsViewModelTest.kt:
```diff
@@ -1,107 +1,166 @@
+        ui = Dispatchers.Unconfined
+        io = Dispatchers.Unconfined
+        background = Dispatchers.Unconfined
     }
```
Can you create a rule with this? Like the `SetMainDispatcherRule`?

victorvicari replied:
Done

faogustavo commented on build.gradle:
```diff
@@ -12,24 +12,28 @@ buildscript {
-    ext.kotlinVersion = "1.3.21"
-    ext.lifecycleVersion = "2.0.0"
+    ext.kotlinVersion = "1.3.41"
+    ext.lifecycleVersion = "2.1.0-rc01"
```
Do you really need the `-rc` version?

victorvicari replied:
It was one of the @ashdavies suggestions

faogustavo reviewed (approved):
I will be nice to create a new issue to migrate all dependencies to a `buildSrc` "module". https://handstandsam.com/2018/02/11/kotlin-buildsrc-for-better-gradle-dependency-management/

victorvicari committed:
updated the tests with unconfined rule. Also added a test for the CollaborationsViewModel

victorvicari committed:
added the Timber call inside the LogException class.
In collaborationViewModel, I replaced from viewModelScope to CoroutineViewModel lauch.

annamorgiel commented:
Hey @victorvicari @faogustavo thanks so much for the input 🌷 How should we process with this one? We could rebase the branch to master and resolve the issues, a fresh PR might be a good idea.

ashdavies commented on app/src/main/java/com/wtmberlin/util/CoroutineViewModel.kt:
```diff
@@ -0,0 +1,27 @@
+var io: CoroutineDispatcher = Dispatchers.IO
+var background: CoroutineDispatcher = Dispatchers.Default
+
+abstract class CoroutineViewModel : ViewModel(), CoroutineScope {
```
This implementation is already available with `viewModelScope` in `lifecycle:2.1.0-RC`

annamorgiel commented:
@victorvicari if you plan to merge the other PR, can this one be closed?

victor-vicari-codeminer42 commented:
Sure thing.

annamorgiel commented:
Closing because update version was merged in #56. Big thanks to everyone!
